pkg/kubectl/cmd/new.go

```diff
@@ -0,0 +1,224 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/spf13/cobra"
+	"io"
+
+	"k8s.io/kubernetes/pkg/api/meta"
+	"k8s.io/kubernetes/pkg/kubectl"
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	"k8s.io/kubernetes/pkg/kubectl/resource"
+	"k8s.io/kubernetes/pkg/runtime"
+)
+
+const (
+	new_long = `Create a new object of the given type using a template.
+
+The new command pulls down an example for the given object type, allows
+you to edit the filled in example, and then submits it for creation.  The
+editing functionality functions similarly to the 'edit' command -- you
+can control the editor used, as well as the format and version, using
+the same flags and environment variables.
+
+By default, before submission your new object will be validated.  If an
+invalid object is encountered, you will be returned to the editor with the
+invalid part indicated with comments.  Use the '--validate=false' in order
+to skip local validation.
+
+You can specify namespace, name, and labels (in 'key1=value1,key2=value2' form)
+from the command line using the '--namespace', '--name', and '--labels' flags.
+
+You can use the '--dry-run' flag to just output the example to standard out
+without launching an editor or submitting the object for creation.
+`
+
+	new_example = `# Create a new pod named 'website':
+  $ kubectl new pod --name website --labels end=front,app=site
+
+  # Don't validate the object before creation
+  $ kubectl new namespace --validate=false --name my-ns
+
+  # Don't edit or create a new object -- just output the example
+  $ kubectl new namespace --name my-ns --dry-run`
+)
+
+func NewCmdNew(f *cmdutil.Factory, out io.Writer) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "new TYPE [--name NAME] [--labels key1=val,key2=val]",
```

Comment:
Because they'd get the same generic behavior?

On Oct 2, 2015, at 2:59 PM, David Eads <notifications@github.com> wrote:

In pkg/kubectl/cmd/new.go
<https://github.com/kubernetes/kubernetes/pull/14294#discussion_r41055691>:

> +without launching an editor or submitting the object for creation.
> +`
> +
> +	new_example = `# Create a new pod named 'website':
> +  $ kubectl new pod --name website --labels end=front,app=site
> +
> +  # Don't validate the object before creation
> +  $ kubectl new namespace --validate=false --name my-ns
> +
> +  # Don't edit or create a new object -- just output the example
> +  $ kubectl new namespace --name my-ns --dry-run`
> +)
> +
> +func NewCmdNew(f *cmdutil.Factory, out io.Writer) *cobra.Command {
> +	cmd := &cobra.Command{
> +		Use:     "new TYPE [--name NAME] [--labels key1=val,key2=val]",

via specific sub-commands -- for instance, in kubectl new rc, rc would be a
subcommand as opposed to an argument.

That makes the help confusing. When you did kubectl new --help, you'd see
some subcommands and you'd also see the available types listed. Some types
would be subcommands and some would be arguments. How would a user know
what to expect?

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/14294/files#r41055691>.